File: src/R/systems/validate_phase14a_integration.R
#!/usr/bin/env Rscript
# Independent base-R validator for Phase 14A.
args <- commandArgs(trailingOnly = TRUE)
root <- if (length(args) > 0) normalizePath(args[[1]], mustWork = TRUE) else normalizePath(".", mustWork = TRUE)

failures <- character()
check <- function(name, condition, detail) {
  if (!isTRUE(condition)) failures <<- c(failures, paste0(name, ": ", detail))
  cat(if (isTRUE(condition)) "PASS " else "FAIL ", name, " — ", detail, "\n", sep = "")
}
path <- function(rel) file.path(root, gsub("/", .Platform$file.sep, rel, fixed = TRUE))
read_csv <- function(rel) read.csv(path(rel), stringsAsFactors = FALSE, check.names = FALSE, na.strings = character())

ontology_rel <- "metadata/atlas_systems.yml"
evidence_rel <- "metadata/atlas_evidence_vocabulary.yml"
identity_rel <- "data/processed/integration/system_identity_crosswalk.csv"
endpoint_rel <- "data/processed/integration/external_endpoint_crosswalk.csv"
relationship_rel <- "data/processed/integration/relationship_taxonomy_inventory.csv"
manifest_rel <- "reports/phase14a_manifest.json"
ontology_lines <- readLines(path(ontology_rel), warn = FALSE, encoding = "UTF-8")
evidence_lines <- readLines(path(evidence_rel), warn = FALSE, encoding = "UTF-8")

system_end <- match("identity_sources:", ontology_lines)
ontology_system_lines <- if (is.na(system_end)) ontology_lines else ontology_lines[seq_len(system_end - 1L)]
system_lines <- ontology_system_lines[grepl("^  - system_id: SYS-", ontology_system_lines)]
system_ids <- sub("^  - system_id: ([^ ]+).*$", "\\1", system_lines)
check("ontology_system_count", length(system_ids) == 13, as.character(length(system_ids)))
check("ontology_system_ids_unique", length(unique(system_ids)) == length(system_ids) && all(grepl("^SYS-", system_ids)), paste(system_ids, collapse = ","))
required_labels <- c("preferred_label:", "atlas_display_label:", "broad_system_family:", "originating_phases:", "canonical_status:", "primary_tables:", "principal_maps:", "temporal_basis:", "baseline_scenario_distinction:", "provenance_pointer:")
check("ontology_required_fields", all(vapply(required_labels, function(x) any(grepl(x, ontology_lines, fixed = TRUE)), logical(1))), paste(required_labels, collapse = ","))

# All listed primary table paths are checked independently from the YAML text.
table_lines <- ontology_lines[grepl("^      - (data/|outputs/)", ontology_lines)]
table_paths <- sub("^      - ", "", table_lines)
missing_tables <- table_paths[!file.exists(path(table_paths))]
check("ontology_listed_paths_exist", length(missing_tables) == 0, paste(missing_tables, collapse = ","))
check("existing_systems_registry_untouched", !file.exists(path(".phase14a_forbidden_systems_marker")), "metadata/systems.yml is not an output of this package")

identity <- read_csv(identity_rel)
endpoint <- read_csv(endpoint_rel)
relationship <- read_csv(relationship_rel)
check("identity_rows", nrow(identity) == 456, as.character(nrow(identity)))
check("identity_atlas_ids_unique", length(unique(identity$atlas_entity_id)) == nrow(identity), as.character(nrow(identity)))
check("identity_source_local_pairs_unique", length(unique(paste(identity$source_artifact, identity$local_id, sep = "|"))) == nrow(identity), as.character(nrow(identity)))
check("identity_mapping_types", all(identity$mapping_type == "exact identity"), paste(unique(identity$mapping_type), collapse = ","))
check("identity_mapping_status", all(identity$mapping_status == "resolved"), paste(unique(identity$mapping_status), collapse = ","))
check("identity_lineage_fields", all(nzchar(identity$source_phase) & nzchar(identity$source_artifact) & nzchar(identity$local_id) & nzchar(identity$evidence_basis)), "blank source lineage")
scenario_rows <- identity[tolower(identity$canon_status) == "scenario" | tolower(identity$reality_status) == "fictional", , drop = FALSE]
scenario_labels_ok <- nrow(scenario_rows) == 0 || all(tolower(scenario_rows$canon_status) == "scenario" & tolower(scenario_rows$reality_status) == "fictional")
check("identity_scenario_labels_retained", scenario_labels_ok, "scenario/fictional identity labels were not retained")

source_id_fields <- c(
  "data/processed/networks/water_system_nodes.csv" = "feature_id",
  "data/processed/networks/materials_system_nodes.csv" = "node_id",
  "data/processed/networks/energy_system_nodes.csv" = "node_id",
  "data/processed/networks/energy_dependency_nodes.csv" = "node_id",
  "data/processed/networks/observation_system_nodes.csv" = "node_id",
  "data/processed/networks/freight_system_nodes.csv" = "node_id",
  "data/processed/networks/ecology_system_nodes.csv" = "node_id",
  "data/processed/networks/exposure_context_nodes.csv" = "node_id",
  "data/processed/networks/biogeochemical_system_nodes.csv" = "node_id",
  "data/processed/networks/climate_hazard_nodes.csv" = "node_id",
  "data/processed/analysis/governance_actors.csv" = "actor_id",
  "data/processed/analysis/governance_authorities.csv" = "authority_id",
  "data/processed/analysis/population_settlement_nodes.csv" = "node_id",
  "data/processed/networks/vector_ecology_nodes.csv" = "node_id",
  "data/processed/networks/infectious_disease_nodes.csv" = "node_id"
)
source_errors <- character()
for (artifact in unique(identity$source_artifact)) {
  if (!(artifact %in% names(source_id_fields))) {
    source_errors <- c(source_errors, paste0("unknown source artifact ", artifact))
  } else {
    d <- read.csv(path(artifact), stringsAsFactors = FALSE, check.names = FALSE, na.strings = character())
    field <- source_id_fields[[artifact]]
    if (!(field %in% names(d))) source_errors <- c(source_errors, paste0("missing field ", artifact, ":", field))
    else if (any(!identity$local_id[identity$source_artifact == artifact] %in% d[[field]])) source_errors <- c(source_errors, paste0("unresolved local ID in ", artifact))
  }
}
check("identity_source_artifacts_and_local_ids", length(source_errors) == 0, paste(source_errors, collapse = "; "))

register <- read_csv("data/processed/analysis/infectious_disease_dependency_register.csv")
conceptual <- grepl("^(EXT|REF)-", register$from_id) | grepl("^(EXT|REF)-", register$to_id)
check("endpoint_rows", nrow(endpoint) == 36, as.character(nrow(endpoint)))
check("endpoint_conceptual_dependency_rows", length(unique(endpoint$dependency_id)) == 25 && all(unique(endpoint$dependency_id) %in% register$dependency_id), as.character(length(unique(endpoint$dependency_id))))
lookup <- match(endpoint$dependency_id, register$dependency_id)
expected <- ifelse(endpoint$endpoint_position == "from", register$from_id[lookup], register$to_id[lookup])
check("endpoint_source_alignment", all(expected == endpoint$local_endpoint_id), "local endpoint differs from Phase 13B register")
status_counts <- table(endpoint$mapping_status)
status_count <- function(name) if (name %in% names(status_counts)) as.integer(status_counts[[name]]) else 0L
check("endpoint_resolution_totals", identical(c(status_count("resolved_system_level"), status_count("retained_conceptual"), status_count("unresolved")), c(34L, 2L, 0L)), paste(names(status_counts), as.integer(status_counts), collapse = ","))
check("endpoint_no_exact_merge", !any(endpoint$mapping_type == "exact identity"), "conceptual endpoint was marked exact")
check("endpoint_unresolved_explicit", all(endpoint$mapping_status != "unresolved" | (endpoint$mapping_type == "unresolved" & endpoint$atlas_entity_id == "")), "unresolved mapping not explicit")
check("endpoint_lineage", all(endpoint$source_phase == "13B" & endpoint$source_artifact == "data/processed/analysis/infectious_disease_dependency_register.csv" & nzchar(endpoint$evidence_basis)), "endpoint provenance incomplete")

required_classes <- c("OBSERVED_DOCUMENTED", "DERIVED_CALCULATED", "INFERRED", "CONTEXT_REUSED", "SCENARIO_ASSUMPTION", "SCENARIO_STATE", "UNRESOLVED", "NONCANONICAL_HOLD")
class_lines <- evidence_lines[grepl("^  - class_id: ", evidence_lines)]
class_ids <- sub("^  - class_id: ([A-Z_]+).*$", "\\1", class_lines)
check("evidence_vocabulary_classes", setequal(class_ids, required_classes), paste(class_ids, collapse = ","))
check("evidence_mapping_count", sum(grepl("^  - local_field: ", evidence_lines)) == 17, as.character(sum(grepl("^  - local_field: ", evidence_lines))))
check("evidence_loss_notes", sum(grepl("information_lost_or_caveat:", evidence_lines, fixed = TRUE)) >= 17, "loss/caveat fields")
required_non_eq <- c("- [fact, inference]", "- [fact, scenario_assumption]", "- [observation, derived_value]", "- [dependency, risk]", "- [association, causation]")
check("evidence_non_equivalences", all(vapply(required_non_eq, function(x) any(grepl(x, evidence_lines, fixed = TRUE)), logical(1))), paste(required_non_eq, collapse = ","))

allowed_classes <- c("physical flow", "material flow", "energy flow", "information / observation", "governance / authority", "operational dependency", "ecological relationship", "exposure pathway", "population / mobility interface", "scenario influence", "high-level interface / association", "unclassified candidate")
check("relationship_rows", nrow(relationship) == 367, as.character(nrow(relationship)))
check("relationship_ids_unique", length(unique(relationship$taxonomy_id)) == nrow(relationship), as.character(nrow(relationship)))
check("relationship_classes", all(relationship$proposed_high_level_class %in% allowed_classes), paste(setdiff(unique(relationship$proposed_high_level_class), allowed_classes), collapse = ","))
check("relationship_inventory_only", all(relationship$normalization_status == "inventory_only_14A"), paste(unique(relationship$normalization_status), collapse = ","))
check("relationship_caveats", all(nzchar(relationship$information_lost_or_caveat)), "blank relationship caveat")
relationship_source_errors <- character()
for (artifact in unique(relationship$source_artifact)) {
  if (!file.exists(path(artifact))) relationship_source_errors <- c(relationship_source_errors, artifact)
}
check("relationship_source_artifacts", length(relationship_source_errors) == 0, paste(relationship_source_errors, collapse = ","))

svg <- path("outputs/figures/western_basin_systems_architecture.svg")
svg_text <- paste(readLines(svg, warn = FALSE, encoding = "UTF-8"), collapse = "\n")
check("architecture_svg", grepl("Western Basin Systems Architecture", svg_text, fixed = TRUE) && grepl("not a geographic map", svg_text, fixed = TRUE), "title or conceptual-map caveat missing")
png <- path("outputs/figures/western_basin_systems_architecture.png")
png_raw <- readBin(png, "raw", n = file.info(png)$size)
png_signature <- as.integer(png_raw[1:8])
check("architecture_png", length(png_raw) > 1000 && identical(png_signature, c(137L, 80L, 78L, 71L, 13L, 10L, 26L, 10L)), paste(length(png_raw), collapse = ","))

sha256_file <- function(file) {
  cmd <- Sys.which("sha256sum")
  if (nchar(cmd) > 0) {
    output <- system2(cmd, file, stdout = TRUE, stderr = TRUE)
    if (length(output) == 0) return(NA_character_)
    return(tolower(gsub("[^0-9a-f]", "", strsplit(trimws(output[[1]]), "[[:space:]]+")[[1]][1])))
  }
  certutil <- Sys.which("certutil")
  if (nchar(certutil) == 0) return(NA_character_)
  output <- system2(certutil, c("-hashfile", file, "SHA256"), stdout = TRUE, stderr = TRUE)
  hits <- tolower(gsub("[[:space:]]+", "", output[grepl("^[0-9A-Fa-f]{64}$", gsub("[[:space:]]+", "", output))]))
  if (length(hits) == 0) NA_character_ else hits[[1]]
}
canonical_text_hash <- function(file) {
  b <- readBin(file, "raw", n = file.info(file)$size)
  # Called only for text extensions; binary frozen artifacts stay raw-byte strict.
  canonical <- charToRaw(gsub("\r\n?", "\n", rawToChar(b), perl = TRUE))
  tmp <- tempfile(fileext = ".txt")
  writeBin(canonical, tmp)
  on.exit(unlink(tmp), add = TRUE)
  sha256_file(tmp)
}
manifest_hash_pairs <- function(file) {
  lines <- readLines(file, warn = FALSE, encoding = "UTF-8")
  current <- NA_character_
  rows <- list()
  for (line in lines) {
    bare_match <- regexec('^[[:space:]]+"([^"]+)"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"[[:space:]]*,?[[:space:]]*$', line, perl = TRUE)
    bare_parts <- regmatches(line, bare_match)[[1]]
    if (length(bare_parts) == 3 && grepl("/", bare_parts[2], fixed = TRUE)) {
      rows[[length(rows) + 1L]] <- c(path = bare_parts[2], sha256 = bare_parts[3])
      current <- NA_character_
      next
    }
    path_match <- regexec('^[[:space:]]+"([^"]+)"[[:space:]]*:', line, perl = TRUE)
    parts <- regmatches(line, path_match)[[1]]
    if (length(parts) == 2 && grepl("/", parts[2], fixed = TRUE)) current <- parts[2]
    hash_match <- regexec('"sha256"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl = TRUE)
    hparts <- regmatches(line, hash_match)[[1]]
    if (length(hparts) == 2 && !is.na(current)) {
      rows[[length(rows) + 1L]] <- c(path = current, sha256 = hparts[2])
      current <- NA_character_
    }
  }
  if (!length(rows)) return(data.frame(path = character(), sha256 = character(), stringsAsFactors = FALSE))
  do.call(rbind, lapply(rows, function(x) data.frame(path = unname(x[["path"]]), sha256 = unname(x[["sha256"]]), stringsAsFactors = FALSE)))
}

freeze_files <- list.files(path("reports"), pattern = "freeze_manifest\\.json$", full.names = TRUE)
freeze_total <- 0L
freeze_unique <- character()
freeze_errors <- character()
for (mf in freeze_files) {
  pairs <- manifest_hash_pairs(mf)
  for (i in seq_len(nrow(pairs))) {
    rel <- pairs$path[[i]]; expected_hash <- pairs$sha256[[i]]; fp <- path(rel)
    freeze_total <- freeze_total + 1L; freeze_unique <- unique(c(freeze_unique, rel))
    if (!file.exists(fp)) freeze_errors <- c(freeze_errors, paste0("missing:", rel))
    else {
      raw_hash <- sha256_file(fp)
      ok <- isTRUE(raw_hash == expected_hash)
      if (!ok && tolower(tools::file_ext(fp)) %in% c("csv", "md", "json", "yml", "yaml", "svg", "txt", "py", "r")) ok <- isTRUE(canonical_text_hash(fp) == expected_hash)
      if (!ok) freeze_errors <- c(freeze_errors, paste0("hash:", rel))
    }
  }
}
check("prior_freeze_integrity", length(freeze_errors) == 0, paste0("entries=", freeze_total, " unique=", length(freeze_unique), " errors=", paste(head(freeze_errors, 5), collapse = ";")))

working_pairs <- manifest_hash_pairs(path(manifest_rel))
working_errors <- character()
for (i in seq_len(nrow(working_pairs))) {
  fp <- path(working_pairs$path[[i]])
  if (!file.exists(fp) || !identical(sha256_file(fp), working_pairs$sha256[[i]])) working_errors <- c(working_errors, working_pairs$path[[i]])
}
check("phase14a_manifest_integrity", length(working_errors) == 0, paste(working_errors, collapse = ","))

status <- system2("git", c("-C", root, "status", "--short"), stdout = TRUE, stderr = TRUE)
forbidden <- status[grepl("metadata/systems\\.yml|glasspunk_base\\.gpkg|freeze_manifest\\.json", status)]
check("no_frozen_or_historical_artifact_modification", length(forbidden) == 0, paste(forbidden, collapse = ";"))
canon <- paste(readLines(path("docs/canon_status.md"), warn = FALSE, encoding = "UTF-8"), collapse = "\n")
check("active_holds_retained", grepl("C — HOLD", canon, fixed = TRUE) && grepl("UNRESOLVED", canon, fixed = TRUE), "hold text")
phase_status <- status[grepl("phase14b|phase15", tolower(status))]
phase_status <- phase_status[!grepl("docs/phase_briefs/phase14b_atlas_layer_registry_cross_system_dependency_normalization.md", phase_status, fixed = TRUE)]
check("phase14b_phase15_not_implemented", length(phase_status) == 0, paste(phase_status, collapse = ";"))

check("all_required_reports", all(file.exists(path(c("reports/phase14a_systems_ontology.md", "reports/phase14a_identity_crosswalk.md", "reports/phase14a_evidence_crosswalk.md", "reports/phase14a_relationship_taxonomy.md", "reports/phase14a_integration_qa.md")))), "report path")

if (length(failures) == 0) {
  cat("PHASE14A_R_VALIDATION PASSED\n")
  quit(status = 0)
} else {
  cat("PHASE14A_R_VALIDATION FAILED\n")
  cat(paste(failures, collapse = "\n"), "\n")
  quit(status = 1)
}

File: src/python/systems/validate_phase14a_integration.py
"""Independent deterministic validator for the Phase 14A integration package."""
from __future__ import annotations

import csv
import hashlib
import json
import re
import subprocess
import sys
import xml.etree.ElementTree as ET
from collections import Counter, defaultdict
from pathlib import Path

import yaml

ROOT = Path(__file__).resolve().parents[3]
ONTOLOGY_PATH = ROOT / "metadata/atlas_systems.yml"
EVIDENCE_PATH = ROOT / "metadata/atlas_evidence_vocabulary.yml"
IDENTITY_PATH = ROOT / "data/processed/integration/system_identity_crosswalk.csv"
ENDPOINT_PATH = ROOT / "data/processed/integration/external_endpoint_crosswalk.csv"
RELATIONSHIP_PATH = ROOT / "data/processed/integration/relationship_taxonomy_inventory.csv"
MANIFEST_PATH = ROOT / "reports/phase14a_manifest.json"
CHECK_PATH = ROOT / "reports/phase14a_artifact_check.json"
TEXT_EXTENSIONS = {".csv", ".md", ".json", ".yml", ".yaml", ".svg", ".txt", ".py", ".r"}
ALLOWED_MAPPING_TYPES = {"exact identity", "system-level concept", "generalized external interface", "unresolved"}
ALLOWED_MAPPING_STATUS = {"resolved", "resolved_system_level", "retained_conceptual", "unresolved"}
ALLOWED_REL_CLASSES = {
    "physical flow", "material flow", "energy flow", "information / observation",
    "governance / authority", "operational dependency", "ecological relationship",
    "exposure pathway", "population / mobility interface", "surveillance / detection",
    "scenario influence", "high-level interface / association", "unclassified candidate",
}


def read_csv(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    with path.open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle)
        return reader.fieldnames or [], list(reader)


def normalize_text(data: bytes) -> bytes:
    return data.replace(b"\r\n", b"\n").replace(b"\r", b"\n")


def sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def hash_matches(path: Path, expected: str) -> tuple[bool, str]:
    raw = path.read_bytes()
    if sha256(raw) == expected:
        return True, "raw"
    if path.suffix.lower() in TEXT_EXTENSIONS and sha256(normalize_text(raw)) == expected:
        return True, "lf_crlf_portable"
    return False, "mismatch"


def manifest_entries(data: dict) -> dict[str, dict]:
    entries = data.get("artifacts", data.get("files", {}))
    return entries if isinstance(entries, dict) else {}


def git_changed_paths() -> list[str]:
    result = subprocess.run(["git", "status", "--short"], cwd=ROOT, text=True, capture_output=True, check=True)
    paths = []
    for line in result.stdout.splitlines():
        if not line.strip():
            continue
        value = line[3:] if len(line) >= 3 else line
        if " -> " in value:
            value = value.split(" -> ", 1)[1]
        paths.append(value.replace("\\", "/"))
    return paths


def main() -> int:
    checks: dict[str, dict[str, object]] = {}
    errors: list[str] = []

    def check(name: str, condition: bool, detail: object) -> None:
        checks[name] = {"passed": bool(condition), "detail": detail}
        if not condition:
            errors.append(f"{name}: {detail}")

    ontology = yaml.safe_load(ONTOLOGY_PATH.read_text(encoding="utf-8"))
    evidence = yaml.safe_load(EVIDENCE_PATH.read_text(encoding="utf-8"))
    systems = ontology.get("systems", [])
    system_ids = [item.get("system_id", "") for item in systems]
    check("ontology_system_count", len(systems) == 13, len(systems))
    check("ontology_system_ids_unique", len(system_ids) == len(set(system_ids)) and all(x.startswith("SYS-") for x in system_ids), system_ids)
    required_system_fields = {"system_id", "preferred_label", "aliases", "broad_system_family", "originating_phases", "canonical_status", "primary_tables", "principal_maps", "temporal_basis", "baseline_scenario_distinction", "provenance_pointer", "atlas_display_label"}
    check("ontology_required_fields", all(required_system_fields.issubset(item) for item in systems), sorted(required_system_fields))
    missing_tables = []
    missing_maps = []
    for item in systems:
        missing_tables.extend([path for path in item["primary_tables"] if not (ROOT / path).exists()])
        for map_name in item["principal_maps"]:
            for suffix in (".png", ".svg"):
                if not (ROOT / "outputs/maps/systems" / f"{map_name}{suffix}").exists():
                    missing_maps.append(f"{map_name}{suffix}")
    check("ontology_primary_tables_exist", not missing_tables, missing_tables[:20])
    check("ontology_principal_maps_exist", not missing_maps, missing_maps[:20])
    check("existing_systems_registry_untouched", not any(p == "metadata/systems.yml" for p in git_changed_paths()), git_changed_paths())

    identity_fields, identity = read_csv(IDENTITY_PATH)
    endpoint_fields, endpoints = read_csv(ENDPOINT_PATH)
    relationship_fields, relationships = read_csv(RELATIONSHIP_PATH)
    check("identity_required_fields", {"atlas_entity_id", "local_id", "source_artifact", "system_id", "mapping_type", "mapping_status", "evidence_basis"}.issubset(identity_fields), identity_fields)
    check("identity_atlas_ids_unique", len({row["atlas_entity_id"] for row in identity}) == len(identity), len(identity))
    check("identity_local_namespace_pairs_unique", len({(row["source_artifact"], row["local_id"]) for row in identity}) == len(identity), len(identity))
    check("identity_mapping_types_valid", all(row["mapping_type"] == "exact identity" for row in identity), sorted({row["mapping_type"] for row in identity}))
    check("identity_mapping_status_valid", all(row["mapping_status"] == "resolved" for row in identity), sorted({row["mapping_status"] for row in identity}))
    check("identity_system_ids_known", all(row["system_id"] in system_ids for row in identity), sorted({row["system_id"] for row in identity if row["system_id"] not in system_ids}))

    source_ids: dict[str, set[str]] = {}
    source_headers: dict[str, list[str]] = {}
    for source in ontology["identity_sources"]:
        fields, records = read_csv(ROOT / source["source_artifact"])
        source_headers[source["source_artifact"]] = fields
        source_ids[source["source_artifact"]] = {record[source["id_field"]].strip() for record in records}
    unresolved_identity = []
    scenario_identity_errors = []
    for row in identity:
        if row["source_artifact"] not in source_ids or row["local_id"] not in source_ids[row["source_artifact"]]:
            unresolved_identity.append(f"{row['source_artifact']}:{row['local_id']}")
        is_scenario = row.get("canon_status", "").lower() == "scenario" or row.get("reality_status", "").lower() == "fictional"
        if is_scenario and not (row.get("canon_status", "").lower() == "scenario" and row.get("reality_status", "").lower() == "fictional"):
            scenario_identity_errors.append(row["atlas_entity_id"])
    check("all_local_ids_resolve_to_source_artifacts", not unresolved_identity, unresolved_identity[:20])
    check("identity_scenario_labels_retained", not scenario_identity_errors, scenario_identity_errors)
    check("identity_provenance_complete", all(row["source_phase"] and row["source_artifact"] and row["local_id"] and row["evidence_basis"] for row in identity), "blank provenance field")
    exact_conflicts = defaultdict(set)
    for row in identity:
        exact_conflicts[(row["source_artifact"], row["local_id"])].add(row["atlas_entity_id"])
    check("no_conflicting_exact_identity_mappings", all(len(values) == 1 for values in exact_conflicts.values()), "conflicting exact identity rows")

    check("endpoint_required_fields", {"dependency_id", "endpoint_position", "local_endpoint_id", "source_artifact", "atlas_entity_id", "mapping_type", "mapping_status", "evidence_basis"}.issubset(endpoint_fields), endpoint_fields)
    register_path = ROOT / "data/processed/analysis/infectious_disease_dependency_register.csv"
    register_fields, register = read_csv(register_path)
    register_by_id = {row["dependency_id"]: row for row in register}
    conceptual_rows = {row["dependency_id"] for row in endpoints}
    expected_conceptual_rows = {row["dependency_id"] for row in register if row["from_id"].startswith(("EXT-", "REF-")) or row["to_id"].startswith(("EXT-", "REF-"))}
    check("endpoint_register_source_exists", register_path.exists(), str(register_path))
    check("endpoint_dependency_ids_resolve", all(row["dependency_id"] in register_by_id for row in endpoints), sorted({row["dependency_id"] for row in endpoints if row["dependency_id"] not in register_by_id}))
    endpoint_mismatch = []
    for row in endpoints:
        source = register_by_id.get(row["dependency_id"], {})
        expected = source.get(f"{row['endpoint_position']}_id")
        if expected != row["local_endpoint_id"]:
            endpoint_mismatch.append(f"{row['endpoint_crosswalk_id']} expected {expected} got {row['local_endpoint_id']}")
    check("endpoint_local_ids_match_source_rows", not endpoint_mismatch, endpoint_mismatch[:20])
    check("endpoint_conceptual_dependency_rows_reconcile", conceptual_rows == expected_conceptual_rows and len(conceptual_rows) == 25, {"actual": len(conceptual_rows), "expected": len(expected_conceptual_rows)})
    endpoint_counts = Counter(row["mapping_status"] for row in endpoints)
    check("endpoint_mapping_status_valid", set(endpoint_counts).issubset({"resolved_system_level", "retained_conceptual", "unresolved"}), dict(endpoint_counts))
    check("endpoint_resolution_totals_reconcile", endpoint_counts == Counter({"resolved_system_level": 34, "retained_conceptual": 2}), dict(endpoint_counts))
    check("endpoint_exact_identity_not_invented", not any(row["mapping_type"] == "exact identity" for row in endpoints), "conceptual endpoint marked exact")
    check("endpoint_unresolved_explicit", all((row["mapping_status"] != "unresolved") or (not row["atlas_entity_id"] and row["mapping_type"] == "unresolved") for row in endpoints), "unresolved endpoint not explicit")
    check("endpoint_system_mappings_known", all((not row["canonical_system_id"]) or row["canonical_system_id"] in system_ids for row in endpoints), sorted({row["canonical_system_id"] for row in endpoints if row["canonical_system_id"] and row["canonical_system_id"] not in system_ids}))
    check("endpoint_provenance_complete", all(row["source_phase"] == "13B" and row["source_artifact"] == "data/processed/analysis/infectious_disease_dependency_register.csv" and row["evidence_basis"] for row in endpoints), "endpoint lineage incomplete")

    normalized_classes = {item["class_id"] for item in evidence.get("normalized_classes", [])}
    required_classes = {"OBSERVED_DOCUMENTED", "DERIVED_CALCULATED", "INFERRED", "CONTEXT_REUSED", "SCENARIO_ASSUMPTION", "SCENARIO_STATE", "UNRESOLVED", "NONCANONICAL_HOLD"}
    check("evidence_vocabulary_complete", normalized_classes == required_classes, sorted(normalized_classes))
    check("evidence_mappings_reference_known_classes", all(item.get("normalized_class") in normalized_classes for item in evidence.get("mappings", [])), sorted({item.get("normalized_class") for item in evidence.get("mappings", []) if item.get("normalized_class") not in normalized_classes}))
    check("evidence_mappings_have_loss_notes", all(item.get("information_lost_or_caveat") for item in evidence.get("mappings", [])), "mapping without information-loss caveat")
    non_eq = {tuple(item) for item in evidence.get("explicit_non_equivalences", [])}
    required_non_eq = {("fact", "inference"), ("fact", "scenario_assumption"), ("fact", "scenario_state"), ("observation", "derived_value"), ("dependency", "risk"), ("association", "causation")}
    check("evidence_non_equivalences_retained", required_non_eq.issubset(non_eq), sorted(required_non_eq - non_eq))
    crosswalk_terms = {(item["local_field"], item["local_term"]) for item in evidence.get("mappings", [])}
    check("evidence_local_terms_are_explicit", all(field and term for field, term in crosswalk_terms), len(crosswalk_terms))

    check("relationship_required_fields", {"taxonomy_id", "source_phase", "source_artifact", "local_field", "local_term", "proposed_high_level_class", "normalization_status", "semantic_compatibility", "information_lost_or_caveat"}.issubset(relationship_fields), relationship_fields)
    check("relationship_taxonomy_ids_unique", len({row["taxonomy_id"] for row in relationships}) == len(relationships), len(relationships))
    relationship_errors = []
    relation_counts = Counter()
    for row in relationships:
        path = ROOT / row["source_artifact"]
        if not path.exists():
            relationship_errors.append(f"missing:{row['source_artifact']}")
            continue
        fields, records = read_csv(path)
        if row["local_field"] not in fields:
            relationship_errors.append(f"field:{row['source_artifact']}:{row['local_field']}")
            continue
        actual_count = sum(1 for record in records if (record.get(row["local_field"]) or "").strip() == row["local_term"])
        if actual_count != int(row["local_term_count"]):
            relationship_errors.append(f"count:{row['taxonomy_id']} expected {actual_count} got {row['local_term_count']}")
        relation_counts[row["proposed_high_level_class"]] += 1
    check("relationship_rows_resolve_and_counts_match", not relationship_errors, relationship_errors[:20])
    check("relationship_classes_allowed", set(relation_counts).issubset(ALLOWED_REL_CLASSES), sorted(set(relation_counts) - ALLOWED_REL_CLASSES))
    check("relationship_inventory_not_claimed_as_14B_normalization", all(row["normalization_status"] == "inventory_only_14A" for row in relationships), sorted({row["normalization_status"] for row in relationships}))
    check("relationship_caveats_present", all(row["information_lost_or_caveat"] for row in relationships), "missing relationship caveat")

    required_reports = [
        "reports/phase14a_systems_ontology.md", "reports/phase14a_identity_crosswalk.md",
        "reports/phase14a_evidence_crosswalk.md", "reports/phase14a_relationship_taxonomy.md",
        "reports/phase14a_integration_qa.md", "reports/phase14a_build_summary.json",
    ]
    check("phase14a_reports_present", all((ROOT / path).exists() and (ROOT / path).stat().st_size > 0 for path in required_reports), required_reports)
    svg = ROOT / "outputs/figures/western_basin_systems_architecture.svg"
    png = ROOT / "outputs/figures/western_basin_systems_architecture.png"
    svg_ok = False
    try:
        ET.parse(svg)
        svg_text = svg.read_text(encoding="utf-8")
        svg_ok = "Western Basin Systems Architecture" in svg_text and "not a geographic map" in svg_text
    except Exception as exc:
        svg_text = str(exc)
    check("architecture_svg_qa", svg_ok, svg_text[:200])
    png_bytes = png.read_bytes() if png.exists() else b""
    png_dims = None
    if len(png_bytes) >= 24 and png_bytes[:8] == b"\x89PNG\r\n\x1a\n":
        png_dims = [int.from_bytes(png_bytes[16:20], "big"), int.from_bytes(png_bytes[20:24], "big")]
    check("architecture_png_qa", bool(png_dims and png_dims[0] >= 1000 and png_dims[1] >= 500), png_dims)

    # Protect all prior freeze-manifest artifacts. Text newline representation is the only tolerated difference.
    freeze_entries = 0
    freeze_unique = set()
    freeze_errors = []
    for manifest_path in sorted((ROOT / "reports").glob("*freeze_manifest.json")):
        data = json.loads(manifest_path.read_text(encoding="utf-8"))
        for rel, meta in manifest_entries(data).items():
            freeze_entries += 1
            freeze_unique.add(rel)
            artifact = ROOT / rel
            expected = meta.get("sha256") if isinstance(meta, dict) else meta
            if not artifact.exists():
                freeze_errors.append(f"missing:{manifest_path.name}:{rel}")
            elif expected:
                ok, mode = hash_matches(artifact, expected)
                if not ok:
                    freeze_errors.append(f"hash:{manifest_path.name}:{rel}")
    check("prior_freeze_integrity", not freeze_errors, {"entries": freeze_entries, "unique_paths": len(freeze_unique), "errors": freeze_errors[:10]})

    manifest_data = json.loads(MANIFEST_PATH.read_text(encoding="utf-8"))
    working_errors = []
    for rel, meta in manifest_entries(manifest_data).items():
        artifact = ROOT / rel
        expected = meta.get("sha256") if isinstance(meta, dict) else meta
        if not artifact.exists():
            working_errors.append(f"missing:{rel}")
        elif expected and not hash_matches(artifact, expected)[0]:
            working_errors.append(f"hash:{rel}")
    check("phase14a_manifest_integrity", not working_errors, {"artifacts": len(manifest_entries(manifest_data)), "errors": working_errors[:10]})

    changed = git_changed_paths()
    forbidden_changes = [path for path in changed if path == "metadata/systems.yml" or path == "data/processed/glasspunk_base.gpkg" or path.startswith("reports/phase") and path.endswith("freeze_manifest.json")]
    check("no_frozen_or_historical_artifact_modification", not forbidden_changes, forbidden_changes)
    check("active_holds_retained", "C — HOLD" in (ROOT / "docs/canon_status.md").read_text(encoding="utf-8") and "UNRESOLVED" in (ROOT / "docs/canon_status.md").read_text(encoding="utf-8"), "active hold text missing")
    forbidden_new = [path for path in changed if path.startswith("data/processed/integration/atlas_layer") or "phase14b" in path.lower() and path not in {"docs/phase_briefs/phase14b_atlas_layer_registry_cross_system_dependency_normalization.md"} or "phase15" in path.lower()]
    check("phase14b_and_phase15_not_implemented", not forbidden_new, forbidden_new)
    check("no_monolithic_geopackage_added", not any(path.endswith(".gpkg") for path in changed), [path for path in changed if path.endswith(".gpkg")])

    counts = {
        "systems": len(systems), "identity_rows": len(identity), "endpoint_rows": len(endpoints),
        "conceptual_dependency_rows": len(conceptual_rows), "total_dependencies": len(register),
        "exact_resolved": 0, "system_level_resolved": endpoint_counts.get("resolved_system_level", 0),
        "retained_conceptual": endpoint_counts.get("retained_conceptual", 0), "unresolved": endpoint_counts.get("unresolved", 0),
        "relationship_rows": len(relationships), "evidence_classes": len(normalized_classes),
        "freeze_manifest_entries": freeze_entries, "freeze_unique_paths": len(freeze_unique),
    }
    result = {"passed": not errors, "checks": checks, "counts": counts, "errors": errors}
    CHECK_PATH.write_text(json.dumps(result, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(json.dumps({"passed": result["passed"], "counts": counts, "errors": errors}, indent=2, sort_keys=True))
    return 0 if result["passed"] else 1


if __name__ == "__main__":
    sys.exit(main())
